### ISSUE TITLE
Bot randomly fails to place blocks when scaffolding

### ISSUE THREAD

**Issue by DEVTomatoCake** [PrismarineJS/mineflayer-pathfinder#296]
scafolding (for search)

When a bot is bridging to a goal and has to place blocks (e.g. when traveling over void) it sometimes randomly fails to place a block but continues to move, which often causes it to fall down (and place blocks mid-air, #181).

I was able to reduce the occurences of this issue by removing
https://github.com/PrismarineJS/mineflayer-pathfinder/blob/56a02b9af715029f884cd392aa7f0836e0284c03/index.js#L500
It seems like nothing is really affected by this change except the bot is bridging slower, which is imo a fair trade off for it to not fall.
Is there another recommended solution for this issue?
Ideal solution would be to fully pause (or find another way) until the block was placed successfully, but it seems like that's not working currently.

**Comment by IceTank** [comment]
That line is suppose to make it fail less. The issue I see with removing this line is that the bot might in some cases not stop crouching.

**Comment by DEVTomatoCake** [comment]
The bot always unsneaked for me when it had to jump - but as I said, another solution is probably better.

**Comment by Muph0** [comment]
I have been digging into this for a while, and turns out it is a timing issue.

The bot triggers a jump here while checking if it is in correct height.
https://github.com/PrismarineJS/mineflayer-pathfinder/blob/56a02b9af715029f884cd392aa7f0836e0284c03/index.js#L486-L489

Then it gets to placing the block like this:
https://github.com/PrismarineJS/mineflayer-pathfinder/blob/56a02b9af715029f884cd392aa7f0836e0284c03/index.js#L490-L497

Two big problems here:
- `bot.equip(block, 'hand')` takes time.
  By the time the block is equipped, the bot might be already gone from the correct position to place it.
- `bot.placeBlock()` takes possibly even more time before the actual block place packet is sent (up to 1 second).

That is because if we look at the implementation of the block placemet, couple of calls down the stack ([`bot.placeBlock()`](https://github.com/PrismarineJS/mineflayer/blob/2e02cee82d6d154b3b7bfb30d213479e7c4fbc59/lib/plugins/place_block.js#L32) --> [`bot._placeBlockWithOptions()`](https://github.com/PrismarineJS/mineflayer/blob/2e02cee82d6d154b3b7bfb30d213479e7c4fbc59/lib/plugins/place_block.js#L6C31-L6C31) --> [`bot._genericPlace()`](https://github.com/PrismarineJS/mineflayer/blob/2e02cee82d6d154b3b7bfb30d213479e7c4fbc59/lib/plugins/generic_place.js#L12)) we can see this, just before the packet is sent:

[plugins/generic_place.js](https://github.com/PrismarineJS/mineflayer/blob/2e02cee82d6d154b3b7bfb30d213479e7c4fbc59/lib/plugins/generic_place.js#L36-L38)
```js
if (options.forceLook !== 'ignore') {
  await bot.lookAt(referenceBlock.position.offset(dx, dy, dz), options.forceLook)
}
```

I'm guessing that the `lookAt` is playing some head movement animation, so we're waiting for this before the packet is sent. By this time, it would be a sheer luck to have the bot in a correct position for the block placemet.

## TLDR
The bot stands in its way when placing a block because it is waiting for too long before sending the place packet. It falls back down and obstructs the block placement.

**Comment by IceTank** [comment]
Yes, equip can take time to select the item, but only when the item is not already in any of the hotbar slots. Placeblock should also send the packet in the same tick if the force look option is trueish. Look will set the bot looking direction and return the promise in the same tick if the force look option is true. Maybe a better approval would be to only start jumping if the item has been equipped. But someone would have to test how well that works.

**Comment by neelnix** [comment]
Upon some debugging, it seems that most of time the bot is unable to place the block due to not getting `true` for this line
https://github.com/PrismarineJS/mineflayer-pathfinder/blob/874949196ac15a7bca23ac54b403b7a8c8bc71be/index.js#L552
Removing the lock fixes most of the issue (_along with the release for it_).

> Yes, equip can take time to select the item, but only when the item is not already in any of the hotbar slots. Placeblock should also send the packet in the same tick if the force look option is trueish. Look will set the bot looking direction and return the promise in the same tick if the force look option is true. Maybe a better approach would be to only start jumping if the item has been equipped. But someone would have to test how well that works.

Tested and found that equipping a scaffolding block is not causing any problem apparently, as with this code for towering, equipping just works fine.

**Comment by BF5258** [comment]
> Removing the lock fixes most of the issue (along with the release for it).

It "fixes" the issue by spamming it until it works. placeBlock isn't spam proofed.

**Comment by BF5258** [comment]
> I'm guessing that the lookAt is playing some head movement animation, so we're waiting for this before the packet is sent.

You are correct. It also doesn't wait until it is above the block in the first place.
Changing line 556 to `bot._placeBlockWithOptions(refBlock, new Vec3(placingBlock.dx, placingBlock.dy, placingBlock.dz), { swingArm: 'right', forceLook: true })` and line 544 to `canPlace = placingBlock.y + 2.1 < bot.entity.position.y`
fixed it for me.
